sonar-kotlin-plugin/src/main/java/org/sonarsource/kotlin/api/FieldMatcher.kt

```diff
@@ -0,0 +1,114 @@
+/*
+ * SonarSource Kotlin
+ * Copyright (C) 2018-2023 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonarsource.kotlin.api
+
+import org.jetbrains.kotlin.descriptors.ClassifierDescriptor
+import org.jetbrains.kotlin.psi.KtNameReferenceExpression
+import org.jetbrains.kotlin.resolve.BindingContext
+import org.jetbrains.kotlin.resolve.descriptorUtil.fqNameSafe
+import org.jetbrains.kotlin.resolve.descriptorUtil.getAllSuperClassifiers
+
+fun interface FieldMatcher {
+    fun matches(node: KtNameReferenceExpression, bindingContext: BindingContext): Boolean
+}
+
+class FieldMatcherBuilderContext {
```

Comment:
I don't think you need KDocs here, names seemed to be pretty self-explanatory.
And we also don't have them for the FunMatcher.

Reply:
I would vote for KDocs here because I think whether an empty set means "match all" or "match none"  is not self-explanatory. If we apply pure set logic, it would be "match none". Since we differ from that, the semantics should be documented.

Reply:
Refactored the `FieldMatcher`. These fields are no more public and the user is forced to rely on the `with*` functions only. Because they are self-explanatory, the problem does no more exist. I removed the comments. Fixed.